### PR TITLE
New annotations and API functions for the decompiler

### PR DESCRIPTION
**Pull request by NirmalManoj**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
As part of the [Improving Decompiler Widget](https://github.com/orgs/radareorg/projects/1) project, many new types of annotations and new API functions have been implemented by the following PRs to the `decompiler-refactoring` branch. This PR is for merging these to the master.
1. #17204 - Annotation for function name 
2. #17281 - Annotations for Constant Variables and Global Variables for the decompiler
3. #17375 - Annotation for local variables
4. #17386 - API for checking if an annotation is a reference or function variable
5. #17397 - Update docs in annotate code API
6. #17403 - Unit tests for annotated code API

...

**Test plan**
1. Go through the linked PRs in order to get an idea of what has been done.
2. Compile, and make sure it compiles properly.
3. Run unit tests. New unit tests are testing all the new API functions and annotations already.
3. Fetch `decompiler-refactoring` branch from the r2ghidra-dec and use the command `pdgj` for the JSON output. See if it's working as expected.

...

**Closing issues**

...

### CODE REVIEW

File: libr/util/annotated_code.c [hunk 1]
```diff
@@ -9,10 +9,28 @@ R_API RAnnotatedCode *r_annotated_code_new(char *code) {
...
+	RCodeAnnotation *annotation = e;
+	if (annotation->type == R_CODE_ANNOTATION_TYPE_FUNCTION_NAME) {
+		free (annotation->reference.name);
+	} else if (annotation->type == R_CODE_ANNOTATION_TYPE_LOCAL_VARIABLE || annotation->type == R_CODE_ANNOTATION_TYPE_FUNCTION_PARAMETER) {
+		free (annotation->variable.name);
+	}
+}
+
+R_API bool r_annotation_is_reference(RCodeAnnotation *annotation) {
+	return (annotation->type == R_CODE_ANNOTATION_TYPE_GLOBAL_VARIABLE || annotation->type == R_CODE_ANNOTATION_TYPE_CONSTANT_VARIABLE || annotation->type == R_CODE_ANNOTATION_TYPE_FUNCTION_NAME);
```
Comment (ITAYC0HEN) [inline libr/util/annotated_code.c:27]: Can you make this LOC shorter by splitting the conditions in new lines? Had to enable line-wrap here

Comment (NirmalManoj) [reply libr/util/annotated_code.c:27]: @ITAYC0HEN Can you tell me how I can do that? When I tried it earlier, the coding style checker was telling me to not split it into multiple lines.

Comment (ITAYC0HEN) [reply libr/util/annotated_code.c:27]: Then I have no idea :)
That's only a small nice-to-have, if it causes a problem then no worries, you can skip this request

**Comment by NirmalManoj** [comment]
I think `openbsd.yml` build is failing due to an unrelated error. Is this the case?

**Comment by XVilka** [comment]
@NirmalManoj yes, it's https://github.com/radareorg/radare2/issues/17395. You can ignore this particular one.

**Comment by ITAYC0HEN** [comment]
Thank you Nirmal! Fantastic work <3

**Pull request merged**